docs/rfcs/20210223_external_connectivity_in_k8s.md

```diff
@@ -0,0 +1,230 @@
+for each Redpanda node. Each instance will
+need to be registered in DNS provider of one's choice.
+
+The `external-dns` project is chosen as a integration point to ease
+the integration with the DNS provider.
+
+## Why (short reason)
+
+The Redpanda is focused on performance and its low tail latency. To
+not add more network hops this proposal tries to merit dynamic
+kubernetes environment with the direct access to Redpanda process.
+
+## How (short plan)
+
+The [PR #658](https://github.com/vectorizedio/redpanda/pull/658)
+introduces Service type NodePort and DNSEndpoint resource. The
+Service resources focus on exposing unique ports in each kubernetes
+node. Operator responsibility is to create statefulset that managed
+Redpanda instances in a form of kubernetes Pod. After statefulset
+creates pods, the operator will map assigned node public IP to
+the unique broker DNS. To achieve that the DNSEndpoint custom
+resource is used which is part of `external-dns` project. It can
+create DNS A records into multiple different DNS providers e.g. `AWS Route53`.
+
+## Impact
+
+Users of the operator should manage `external-dns` by their own.
+The integration point is the DNSEndpoint custom resource which
+`external-dns` needs to start watching it. The kubernetes worker
+nodes need open port per Redpanda cluster which might be an
+additional attack vector.
+
+# Motivation
+## Why are we doing this?
+
+Redpanda can be easily managed in multiple environment, but vast
+range of users plan to manage their workload in kubernetes
+environment. Saying that, teams would like to not introduce
+clients application into their kubernetes clusters. The Kafka API
+requires that each broker will be reachable at any given time.
+Canonical way of exposing webservers is to create Ingress resources
+that can be reconciled into cloud provider load balancer. Another
+way of exposing different kind of workloads is to expose their TCP
+port using again cloud provider load balancer. Ont the other hand
+the TCP port can be exposed on each kubernetes worker node using
+Service type Node Port. All the mentioned methods doesn't guarantee
+that each Redpanda node will be individually accessible. For that
+reasons the kubernetes model of exposing workloads outside of the
+k8s cluster doesn't provide predictable way of reaching individual
+broker in Redpanda cluster.
+
+There must be easily identifiable Redpanda node address (DNS).
+Those addresses should accommodate dynamic nature of any cloud
+where nodes can disappear or be rescheduled.
+
+## What use cases does it support?
+
+Exposing each Redpanda node to the internet.
+
+## What is the expected outcome?
+
+Users of the Redpanda operator can easily integrate with the cloud
+providers DNS service. The nodes should be reachable from the
+internet after the reconciliation ends. With the caveat that DNS
+records can propagate up to 24-48 hours.
```

Comment:
We have ways to bring that time down, eg having our records accompanied with a small ttl. Although not everyone on the internet respects those :)

Reply:
Agree. Never the less I will remove the DNS part from the external connectivity feature.